src/plugins/links/public/embeddable/links_embeddable.tsx

```diff
@@ -6,152 +6,228 @@
+import { ExternalLinkComponent } from '../components/external_link/external_link_component';
+import {
+  LinksApi,
+  LinksByReferenceSerializedState,
+  LinksByValueSerializedState,
+  LinksRuntimeState,
+  LinksSerializedState,
+} from '../types';
+import { APP_NAME } from '../../common';
+import { initializeLinks } from './initialize_links';
+import { injectReferences } from '../../common/persistable_state';
+import { openEditorFlyout } from '../editor/open_editor_flyout';
+
+import '../components/links_component.scss';
+import { checkForDuplicateTitle, linksClient } from '../content_management';
+import { resolveLinks } from '../lib/resolve_links';
+import {
+  deserializeLinksSavedObject,
+  linksSerializeStateIsByReference,
+} from '../lib/deserialize_from_library';
+
+export const LinksContext = createContext<LinksApi | null>(null);
+
+export const getLinksEmbeddableFactory = () => {
+  const linksEmbeddableFactory: ReactEmbeddableFactory<
+    LinksSerializedState,
+    LinksRuntimeState,
+    LinksApi
+  > = {
+    type: CONTENT_ID,
+    deserializeState: async (serializedState) => {
+      // Clone the state to avoid an object not extensible error when injecting references
+      const state = cloneDeep(serializedState.rawState);
+      const { title, description } = serializedState.rawState;
+
+      if (linksSerializeStateIsByReference(state)) {
+        const attributes = await deserializeLinksSavedObject(state);
+        return {
+          ...attributes,
+          title,
+          description,
+        };
+      }
+
+      const { attributes: attributesWithInjectedIds } = injectReferences({
+        attributes: state.attributes,
+        references: serializedState.references ?? [],
+      });
+
+      const resolvedLinks = await resolveLinks(attributesWithInjectedIds.links ?? []);
+
+      return {
+        title,
+        description,
+        links: resolvedLinks,
+        layout: attributesWithInjectedIds.layout,
+        defaultPanelTitle: attributesWithInjectedIds.title,
+        defaultPanelDescription: attributesWithInjectedIds.description,
+      };
+    },
+    buildEmbeddable: async (state, buildApi, uuid, parentApi) => {
+      const { titlesApi, titleComparators, serializeTitles } = initializeTitles(state);
+      const { linksApi, linksComparators, serializeLinks } = initializeLinks(state, parentApi);
+      const savedObjectId$ = new BehaviorSubject(state.savedObjectId);
+
+      const serializeState = async (): Promise<SerializedPanelState<LinksSerializedState>> => {
+        const { attributes, references } = serializeLinks();
+        if (savedObjectId$.value !== undefined) {
+          const linksByReferenceState: LinksByReferenceSerializedState = {
+            savedObjectId: savedObjectId$.value,
+            ...serializeTitles(),
+          };
+
+          return { rawState: linksByReferenceState, references: [] };
+        }
```

Comment:
I wonder - should we follow the pattern of the book example embeddable where the saved object is updated **when the dashboard is saved** rather than when the editor is closed? Right now, it's a bit confusing that editing a by-reference panel triggers unsaved changes on the dashboard, despite the fact that the saved object was already updated...

Reply:
Whichever way we handle this, we should be consistent. Currently, inline editing a by-reference Lens embeddable saves the changes to the Lens saved object when the flyout is closed by clicking "Apply". AFAICT, Lens and Links are the only two by-reference embeddables that can be edited inline. 

It looks like @dej611 is [keeping the existing](https://github.com/elastic/kibana/pull/186642/files#diff-4902324e9656eeb92fea70683b9b2013bf67a5a5b3d37bf9291575902dea40eaL132) functionality of updating the saved object on flyout close. So I wonder if we should wait on this suggestion with the potential to unify the inline-editing pattern later in a single PR?

cc @nreese

Reply:
> Currently, inline editing a by-reference Lens embeddable saves the changes to the Lens saved object when the flyout is closed by clicking "Apply". AFAICT, Lens and Links are the only two by-reference embeddables that can be edited inline.

Agreed that we should be consistent! I'm fine if we update the saved object when the flyout is closed, but in that case, unsaved changes **should not be triggered** when the panel being updated is by-reference (this is how it worked with the links panel **before** this refactor):


https://github.com/elastic/kibana/assets/8698078/9bc1ec8f-5288-46f4-8354-9c7abba4fc24



I tested with Lens, and unfortunately they are triggering unsaved changes even when the panel is by-reference - but this is confusing to me, because at that point, the "Reset" button **doesn't actually do anything** when a by-reference panel is edited. You click it and.... nothing happens. Really not great UX IMO.

Reply:
Good point @Heenawter! I don't see a good way to avoid the dirty state when saving on flyout close. I'll see if I can change the code to save the by-reference changes on dashboard save instead.

Reply:
I've tried to implement the suggestion to save by-ref panels when dashboard saves, but now I'm feeling like the changes required is out of scope for this PR. Instead, I've updated the comparators to ignore certain attribute changes when the panel is by-reference in https://github.com/elastic/kibana/pull/178670/commits/3084f31eca0552b70b44716d6663e17998bd6926.

I'll publish my WIP on the save by-ref panels on dashboard save in a separate branch so we can pick this up again later if we want.